### DISCUSSION TITLE
Adding TLS 1.3 Early Data Support

### DISCUSSION THREAD

**Discussion by LaRuim** [curl/curl#13743] [Ideas]
I plan on mentoring a summer project with a small team of undergrad students to add early data support to cURL over about 8 weeks starting in June. (https://github.com/curl/curl/discussions/13528)

I'd made a basic 8-week ish plan to help facilitate this smoothly. I wanted the community's input on it, as well as the preferred method of contribution (the entire feature in one PR, or split over multiple ones); changes within the PR(s) will be  logically split across commits as the guidelines say.

**Comment by LaRuim** [comment]
### **Roadmap**

**Week 1:** Background reading. What is TLS? What are sessions and session tickets? What is a PSK and why does it matter? What exactly is 0-RTT? What is Session Resumption? What is Early Data? (This is conservative, the idea is that the team knows some basics already.)

**Week 2:** Use openssl s_client and Wireshark to figure out how early data looks from a packet PoV and intuitively understand it. Explore the cURL codebase; understand how it interacts with the underlying SSL layer (wolfssl or openssl); track handshake and request flow, look at how curl does tlsv1.3 session resumption; look at how curl supports multiple requests over the same connection (if at all.)

**Weeks 3-6:** Start work on implementation; 
Checkpoints include using an early data identifier, check for supported methods only, retrieving session parameters and using openssl APIs to write early data, handling a 'Too Early' reply and retrying the same request post-handshake, switching over to new keys once the handshake is done.

**Week 7:** Write unit tests, polish code, add documentation.

**Week 8 (and beyond, if necessary):** Open the PR and solicit feedback, and work on addressing feedback

**Reply by LaRuim** [reply]
This is a rough plan; I completely expect there to be hiccups along the way and I'm alright the timelines extending or shifting.

**Comment by bagder** [comment]
Awesome. Welcome, and I wish you the best of luck with this awesome project!

I can't really comment on how feasible this schedule is. It depends so much on the specific persons involved, their background, experiences and prior knowledge.

## Week 1

A challenge with everything TLS in curl is that we have an internal API for TLS operations and we have many different backends that all at least need to build. It is fine to add features for one or a specific subset of TLS backends, but bonus points is of course given the more backends you can cover.

## Week 2

Yes, curl does multiplexing (multiple requests over the same connection) over both HTTP/2 and HTTP/3.

## Week 7

A more modern approach is to create a set of basic tests earlier in the project to have something to work against in more controlled way. I think maybe planning for and thinking about how to tests can and should be done is probably something to do as some of the first steps in this project.

Also: if there is no documentation written for the functionality until week 7, then it will be hard to review and comment on the functionality and design until then, which then increases the risk that the solution and architecture ends up sub-optimal or a larger retake might be required. I think at least a first/rough version of the documentation for how the feature is supposed to be used by users need to be provided early on, as it might affect the implementation.

## Approach

If the feature gets implemented as a set of independent features/options, then they should be separate PRs. It does however sound like this would be a single one, which then would be most suitable as a single PR. 

I would then propose to file a draft PR as early as possible to attract feedback and comments on the chosen solution. Further work could then just rebase that branch and force-push when there are updates. It will also help make sure that no unrelated builds break because of the work etc.

**Comment by LaRuim** [comment]
That's extremely usedul feedback, thank you!

I think it absolutely makes sense to write a few basic tests in the beginning, and I'll edit the plan accordingly; The documentation part too makes sense!

Based on this, I'll go with a less time-based PR opening approach; we'll file one as soon as we've got something working, hopefully that's earlier on in the process.

What I'm also realising is that the process may be slightly different for HTTP/2, and quite different for HTTP/3. I shall also keep these in mind. Perhaps separate PRs per protocol could also work?

I will be pitching this as a project to a set of approximately 25 students; this is part of a little (senior) student and alumni driven program where multiple such projects are offered, so all of this is contingent on 2-4 (junior) students actually choosing to work on this; just a heads up!

If the project finds enough suitable students to work on it, there will be bi-weekly progress presentation meetings by the team to a wider audience; is this perhaps something you'd be interested in being a part of?

Once again, thank you!

**Comment by LaRuim** [comment]
We've opened a [PR here.](https://github.com/curl/curl/pull/14135/commits/9248e70a7dc4b32ac00766ffa28993135126ea17)
